matcher/cache/cache.go

```diff
@@ -155,34 +165,55 @@ func (c *cache) ForwardMatch(namespace, id []byte, fromNanos, toNanos int64) rul
 
 func (c *cache) Register(namespace []byte, source matcher.Source) {
 	c.Lock()
+	defer c.Unlock()
+
+	nsHash := xid.HashFn(namespace)
+	if results, exist := c.namespaces[nsHash]; !exist {
+		c.namespaces[nsHash] = newResults(source)
+		c.metrics.registers.Inc(1)
+	} else {
+		c.updateWithLock(nsHash, source, results)
+		c.metrics.registerExists.Inc(1)
+	}
+}
+
+func (c *cache) Update(namespace []byte, source matcher.Source) {
+	c.Lock()
+	defer c.Unlock()
+
 	nsHash := xid.HashFn(namespace)
 	results, exist := c.namespaces[nsHash]
+	// NB: The namespace does not exist yet. This could happen if the source update came
+	// before its namespace is registered. It is safe to ignore this premature update
+	// because the namespace will eventually register itself and refreshes the cache.
 	if !exist {
-		results = newResults(source)
-	} else {
-		// Invalidate existing cached results.
-		c.toDelete = append(c.toDelete, results.elems)
-		c.notifyDeletion()
-		results.source = source
-		results.elems = make(elemMap)
+		c.metrics.updateNotExists.Inc(1)
+		return
 	}
-	c.namespaces[nsHash] = results
-	c.Unlock()
-	c.metrics.registers.Inc(1)
+	// NB: The source to update is different from what's stored in the cache. This could
+	// happen if the namespace is changed, removed, and then revived before the rule change
+	// could be processed. It is safe to ignore this stale update because the last rule
+	// change update will eventually be processed and the cache will be refreshed.
+	if results.source != source {
+		c.metrics.updateStaleSource.Inc(1)
+		return
+	}
+	c.updateWithLock(nsHash, source, results)
```

Comment:
make a separate function "cleanUpCache" is probably more clear since you are not updating the source, and you specifically only want to do this when the source remain unchanged.

Reply:
Sure, would `refreshWithLock` be a better name?

Reply:
sg, yeah "refresh" sounds better